src/get-started/test-service.md

```diff
@@ -0,0 +1,365 @@
+    assert_eq!(wallet.balance(), 100);
+}
+```
+
+Per Rust conventions, the test is implemented as a zero-argument function
+without the returned value and
+with a `#[test]` annotation. This function will be invoked during testing;
+if it does not panic, the test is considered passed.
+
+We use one of `create_block*` methods defined by `TestKit` to send a transaction
+to the testkit node and create a block with it (and only it). Then, we use
+the service schema to check that the transaction has led to
+the expected changes in the storage.
+
+To run the test, execute `cargo test` in the shell:
+
+```none
+$ cargo test
+# (Some output skipped)
+running 1 test
+test test_create_wallet ... ok
+
+test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+```
+
+If we remove or comment out the `create_block*` call, the test
+will expectedly fail because the wallet is no longer created:
+
+```none
+running 1 test
+test test_create_wallet ... FAILED
+
+failures:
+
+---- test_create_wallet stdout ----
+  thread 'test_create_wallet' panicked at 'No wallet persisted'
+```
+
+### Successful Transfer
+
+> **Test:** `test_transfer`
+
+Let’s test a transfer between two wallets. First, we need to create the testkit
+and the wallets:
+
+```rust
+let mut testkit = init_testkit();
+let (alice_pubkey, alice_key) = crypto::gen_keypair();
+let (bob_pubkey, bob_key) = crypto::gen_keypair();
+testkit.create_block_with_transactions(txvec![
+    TxCreateWallet::new(&alice_pubkey, "Alice", &alice_key),
+    TxCreateWallet::new(&bob_pubkey, "Bob", &bob_key),
+    TxTransfer::new(
+        &alice_pubkey, // sender
+        &bob_pubkey, // receiver
+        10, // amount
+        0, // seed
+        &alice_key, // private key used to sign the transaction
+    ),
+]);
+```
+
+Check that wallets are committed to the blockchain and have expected balances:
+
+```rust
+let wallets = {
+    let snapshot = testkit.snapshot();
+    let schema = CurrencySchema::new(&snapshot);
+    (schema.wallet(&alice_pubkey), schema.wallet(&bob_pubkey))
+};
+if let (Some(alice_wallet), Some(bob_wallet)) = wallets {
+    assert_eq!(alice_wallet.balance(), 90);
+    assert_eq!(bob_wallet.balance(), 110);
+} else {
+    panic!("Wallets not persisted");
+}
+```
+
+### Transfer to Non-Existing Wallet
+
+> **Test:** `test_transfer_to_nonexisting_wallet`
+
+Unlike in real Exonum network, you can control which transactions the testkit
+will include into the next block. This allows to test different orderings
+of transactions, even those that would be hard (but not impossible) to reproduce
+in the real network.
+
+Let’s test a case when Alice sends a transaction to Bob while the Bob’s wallet
+is not committed. The test is quite similar to the previous one, with the exception
+how the created transactions are placed into the block.
+Namely, the `create_block_with_transactions` call is replaced with
+
+```rust
+testkit.create_block_with_transactions(txvec![
```

Comment:
I thought we doesn't have confidence that transactions are executed one-by-one. Am I right? We need to split this block in three to make sure Bob's wallet will be created *before* the transfer.

Reply:
> I thought we don't have confidence that transactions are executed one-by-one. Am I right?

In general, Exonum executes transactions sequentially, one-by-one in the order they are listed in a block. If transactions could be executed in parallel, this could lead to undefined behavior and consensus loss. (Certain concurrency models could alleviate this, but they require *a lot* of effort and are not implemented in Exonum right now.)

As for testkit specifically, `create_block_with_transactions` creates a block with the transactions listed in the exact specified order. So, it's safe to assume that Bob's wallet will be created before the transfer.